Enhance version handling in packaging (New)

fernando79513 opened the pull request:
<!--
Example Title: Fixed bugged behavior of checkbox load config (Bugfix)

A Traceability Marker is required as a suffix in the PR title to help understand the impact of your change at a glance.

Pick one of the following:
- Infra: Your change only includes documentation, comments, GitHub actions or meta box
- BugFix: Your change fixes a bug
- New: Your change is a new backward-compatible feature, a new test/test plan/test inclusion
- Breaking: Your change breaks backward compatibility.
    - This includes any API change to checkbox-ng/checkbox-support
    - Changes to PXU grammar/field requirements
    - Breaking changes to dependencies in snaps (fits upgrade for example)

If your change is to providers it can only be (Infra, BugFix or New).

If your change impacts the submission format in Checkbox test reports, ensure that `submission-schema/schema.json` is updated and relevant fields are documented.

Signed commits are required.
  - See CONTRIBUTING.md (https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#signed-commits-required) for further instructions.
  - If you are posting your first pull request from a fork of the repository, a Checkbox maintainer (someone with contributor / maintainer / admin rights) will be required to enable CI checks in the repo to be executed.
    - This will be communicated with a comment to the PR of the form `/canonical/self-hosted-runners/run-workflows <SHA-for-HEAD-commit>`
-->

## Description

<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
-->

These changes introduce a way to use comparisons for PPA packaging. Now we can use expressions like ">=22.04" in `packaging.pxu` files:

```
unit: packaging meta-data
os-id: ubuntu
os-version-id: >=22.04
Depends: libsvm3
```

This allows to use package dependencies for multiple ubuntu versions without having to declare them individually.

## Resolved issues

<!--
Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
Make sure that the linked issue titles & descriptions are also up to date.
-->

Closes [CHECKBOX-1062](https://warthogs.atlassian.net/browse/CHECKBOX-1062)

## Documentation

<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Documentation in the repository, including contribution guidelines.
  - Process documentation outside the repository.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
-->

The documentation of "Packaging Meta Data Unit" has been updated with this new change.
## Tests

<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
- Please provide a list of what tests were run and on what platform/configuration.
- Remember to check the test coverage of your PR as described in CONTRIBUTING.md
-->

There are two tests, one that tests the functionality directly and a higher-level one that uses metabox to ensure that the packaging is properly compared for a specific ubuntu version.

To run the packaging tests:
```
cd ~/checkbox/checkbox-ng
python3 -m pytest -k DebianPackagingDriverTests
```

For the Metabox tests, you can run:
 - `metabox configs/local-source-18.04.py --tag packaging`: the test should not pass in bionic
 - `metabox configs/local-source-20.04.py --tag packaging`: the test should pass in focal
 - `metabox configs/local-source-22.04.py --tag packaging`: the test should pass in jammy

It is also possible to test manually that the package dependencies are created properly. If you run:
```
cd ~/checkbox/providers/base
./manage.py packaging
cat debian/checkbox-provider-base.substvars
```

You should see `libsvm3` and `python3-opencv` added for version 20.04 or superior, and not for earlier versions



[CHECKBOX-1062]: https://warthogs.atlassian.net/browse/CHECKBOX-1062?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kissiel reviewed (changes requested):
This type of code is a an absolute joy to write unit tests for: simple input ->simple output.

The metabox scenarios provided here are testing something that should be tested with unit tests.

One general question would be how to deal with compound expressions, or multiple expressions resulting in conflicting results. All of those I think could/should be covered by UTs.

I also wrote some comments below.

fernando79513 commented on checkbox-ng/plainbox/impl/unit/packaging.py:
```diff
@@ -334,22 +305,81 @@ def __init__(self, os_release: 'Dict[str, str]'):
+        # Default to '==' if no operator is provided
+        operators = {
+            None: operator.eq,
+            '=': operator.eq,
```
To be sure, you are proposing to remove `'=': operator.eq,` in favor of only using `==`, right?

fernando79513 replied:
Done!

fernando79513 commented on checkbox-ng/plainbox/impl/unit/packaging.py:
```diff
@@ -334,22 +305,81 @@ def __init__(self, os_release: 'Dict[str, str]'):
+            r"(=|==|>=|<=|>|<|!=)?\s*([\d\.a-zA-Z]+)", comparison_string
+        )
+        if not match:
+            raise ValueError("Invalid version comparison string")
```
I think that this piece of code is only called when running "manage.py packaging" for a provider. This script calls the `invoked()` method in the `PackagingCommand` class, which calls `inspect_provider` which calls to the comparision functionality.

So if we have a misformed version requirement string, it will only crash when running the `manage.py packaging` script, which I think makes sense, since you don't want to create the packaging file with a malformed string.

That's at least my opinion, I don't know If I am missing something.

kissiel replied:
That's sounds good. The followup:
* let's not make it print the stacktrace - this is not program's fault that the input is wrong -> raise SystemExit(...)
* let's add context -> let's quote the string that's bad

fernando79513 replied:
Done!

fernando79513 commented:
> This type of code is a an absolute joy to write unit tests for: simple input ->simple output.
> 
> The metabox scenarios provided here are testing something that should be tested with unit tests.
> 
> One general question would be how to deal with compound expressions, or multiple expressions resulting in conflicting results. All of those I think could/should be covered by UTs.
> 
> I also wrote some comments below.

I haven't added the possibility of interpreting more than one comparison per packaging metadata unit, so compounded expressions are not contemplated. It should be feasible to check multiple expressions, but that was not required by any package at the moment.

If you think that handling multiple expressions at the same time is going to be needed in the future, I can implement that functionality too and add some tests for that.

kissiel commented on metabox/metabox/metabox-provider/bin/compare_packaging_version.py:
```diff
@@ -0,0 +1,53 @@
+    return 1
+
+
+def main(argv: str):
```
It's not only about the approach, it's about what is being tested and applying the simplest/smallest tool for the job. If something can be tested within the python itself (like the functions here) it should be a unit or an integration test (AKA pytest test).

If the test recreates a real-world use case, it probably needs a "real" system, with its version and everything, and this is where metabox fits perfectly.

fernando79513 replied:
Done!

kissiel commented:
> > This type of code is a an absolute joy to write unit tests for: simple input ->simple output.
> > The metabox scenarios provided here are testing something that should be tested with unit tests.
> > One general question would be how to deal with compound expressions, or multiple expressions resulting in conflicting results. All of those I think could/should be covered by UTs.
> > I also wrote some comments below.
> 
> I haven't added the possibility of interpreting more than one comparison per packaging metadata unit, so compounded expressions are not contemplated. It should be feasible to check multiple expressions, but that was not required by any package at the moment.
> 
> If you think that handling multiple expressions at the same time is going to be needed in the future, I can implement that functionality too and add some tests for that.

I don't want to support it, on the contrary  - the simpler the better. But the question "what happens when someone declares two similar pieces of packaging metadata that yield different results" needs to be answered (and possibly reflected in tests).

yphus commented:
1. I second Maciej on adopting a unit test approach rather than new metabox scenarios.

2. Regarding units yielding different results, I'd also prefer to keep things simple. It means a package should only be listed once and having packaging meta data units with the same package name listed should fail validation.
I don't think we have to support situations where a package would be required for e.g  <18.04 and >20.04 (which would imply two units to express the two ranges)
So I'd advocate for a new validator that checks the content of the Depends/Recommends/Suggests fields of a unit (splitting the content since it can be a list). For each package listed run a validation in context to be sure there's only one unit that references it (for maintainability)

fernando79513 commented:
> 1. I second Maciej on adopting a unit test approach rather than new metabox scenarios.
> 2. Regarding units yielding different results, I'd also prefer to keep things simple. It means a package should only be listed once and having packaging meta data units with the same package name listed should fail validation.
>    I don't think we have to support situations where a package would be required for e.g  <18.04 and >20.04 (which would imply two units to express the two ranges)
>    So I'd advocate for a new validator that checks the content of the Depends/Recommends/Suggests fields of a unit (splitting the content since it can be a list). For each package listed run a validation in context to be sure there's only one unit that references it (for maintainability)

As Maciej suggested during the SU, I will add the validation checking for repeated packages in another PR, so we can close this one and don't make it even bigger.

kissiel reviewed (changes requested):
The metabox tests still miss the target. Please exclude them from this PR, and file a ticket on Jira to add metabox coverage for packaging units, and schedule some pairing session so we may collaborate on this. After the MB is removed from this PR, we can land it.

fernando79513 commented:
> The metabox tests still miss the target. Please exclude them from this PR, and file a ticket on Jira to add metabox coverage for packaging units, and schedule some pairing session so we may collaborate on this. After the MB is removed from this PR, we can land it.

I have removed the metabox tests and added them to a new branch. 
I have also created a ticket to address these test: https://warthogs.atlassian.net/browse/CHECKBOX-1156

kissiel reviewed (approved):
Fantastic set of changes!
TYVM!
